homeassistant/components/axis/camera.py

```diff
@@ -35,52 +35,71 @@ class AxisCamera(AxisEntity, MjpegCamera):
 
     _attr_supported_features = CameraEntityFeature.STREAM
 
+    _still_image_url: str
+    _mjpeg_url: str
+    _stream_source: str
+
     def __init__(self, device: AxisNetworkDevice) -> None:
         """Initialize Axis Communications camera component."""
         AxisEntity.__init__(self, device)
 
+        self._generate_sources()
+
         MjpegCamera.__init__(
             self,
             username=device.username,
             password=device.password,
             mjpeg_url=self.mjpeg_source,
             still_image_url=self.image_source,
             authentication=HTTP_DIGEST_AUTHENTICATION,
+            unique_id=f"{device.unique_id}-camera",
         )
 
-        self._attr_unique_id = f"{device.unique_id}-camera"
-
     async def async_added_to_hass(self) -> None:
         """Subscribe camera events."""
         self.async_on_remove(
             async_dispatcher_connect(
-                self.hass, self.device.signal_new_address, self._new_address
+                self.hass, self.device.signal_new_address, self._generate_sources
             )
         )
 
         await super().async_added_to_hass()
 
-    def _new_address(self) -> None:
-        """Set new device address for video stream."""
-        self._mjpeg_url = self.mjpeg_source
-        self._still_image_url = self.image_source
+    def _generate_sources(self) -> None:
+        """Generate sources.
+
+        Additionally used when device change IP address.
+        """
+        image_options = self.generate_options(skip_stream_profile=True)
+        self._still_image_url = f"http://{self.device.host}:{self.device.port}/axis-cgi/jpg/image.cgi{image_options}"
+
+        mjpeg_options = self.generate_options()
+        self._mjpeg_url = f"http://{self.device.host}:{self.device.port}/axis-cgi/mjpg/video.cgi{mjpeg_options}"
+
+        stream_options = self.generate_options(add_video_codec_h264=True)
+        self._stream_source = f"rtsp://{self.device.username}:{self.device.password}@{self.device.host}/axis-media/media.amp{stream_options}"
+
+        self.device.additional_diagnostics["camera_sources"] = {
+            "Image": self._still_image_url,
+            "MJPEG": self._mjpeg_url,
+            "Stream": self._stream_source.replace(self.device.username, "user").replace(
+                self.device.password, "pass"
+            ),
```

Comment:
Why rely on replacements if you can just set it without?

```suggestion
            "Stream": f"rtsp://user:pass@{self.device.host}/axis-media/media.amp{stream_options}",
```

Reply:
There will be more variations of that url in the future so better to change it than duplicating it

Reply:
Right, but replacing usernames/password is, IMHO, bad practice (easy to mess up). It is better to have them not there in the first place at all...

Reply:
I'll follow your suggestion and rethink it once another source is in play :)